# Importance of Analyzing Supermarket Sales Data for Business Insights
**When Data Analysis Reveals Hidden Wizardry:**

![Data Meme](https://th.bing.com/th/id/OIP.j1B9I8aMy36jdc5tQbIVNwHaF8?rs=1&pid=ImgDetMain)

> **"Uncover the unexpected with data analysis – because insights can be as delightful as a good spell!"**


## 1. Performance Evaluation

- **Product Vibes:** Unleash statistical kung fu to spot rockstar products and the underachievers, paving the way for inventory magic.
- **Branch Drama:** Wield data visualization spells to rate each branch's performance, helping them level up.

## 2. Customer Shenanigans

- **Clairvoyant Clustering:** Use ML wizardry for sorting customers into Hogwarts-like houses based on their buying spells.
- **Preference Potions:** Stir statistical cauldrons to reveal customer preferences for products and payment spells.

## 3. Operational Wizardry

- **Inventory Alchemy:** Apply data preprocessing spells to scrub and prepare data for inventory enchantments.
- **Supply Chain Spells:** Use Python sorcery to decode sales patterns and optimize the supply chain spellbook.

## 4. Strategic Sorcery

- **Trend Divination:** Use statistical spells to predict market trends, and sprinkle ML fairy dust for a glimpse into the retail crystal ball.
- **Competitive Charms:** Employ Python charms for competitive analysis, outshining industry benchmarks.

## 5. Marketing Magic

- **Campaign Wizardry:** Use Python spells to rate the success of marketing campaigns, ensuring they are more magic than muggle.
- **Seasonal Sorcery:** Apply statistical charms to reveal seasonal swings for targeted promotions.

## 6. Financial Wizardry

- **Revenue Runes:** Use ML sorcery for predicting revenue futures, because who needs a crystal ball?
- **Cost Conundrums:** Employ statistical enchantments to analyze the cost of wizardry and gross income magic.

## 7. Customer Experience Spells

- **Rating Rituals:** Use statistical rituals to align customer satisfaction ratings with sales magic.
- **Time-Turner Analysis:** Utilize Python time-turners to analyze sales patterns during specific hours for optimal staff summoning.

> **In conclusion**,the retail enchantress, your mix of statistical alchemy, Python sorcery, and machine learning magic can turn businesses into retail wizards—ready to charm their way to success.

# Supermarket Sales Data Analysis Overview

![Supermarket Sales Data](https://static.businessinsider.com/image/525ec3f1ecad040036ec7640/image.jpg)


The supermarket sales dataset unveils the magical sales history of a prominent retail wizarding chain. Recorded over three enchanted months and spanning three magical branches (A, B, and C), this dataset is a treasure trove for predictive data enchantments. It provides insights into customer spells, product charms, and the overall wizarding business trends.

## Dataset Overview

- **Invoice ID:** Magical identification for each enchanted sales slip.
- **Branch:** Wizarding branches labeled A, B, and C.
- **City:** Locations where the magic happens, reflecting the diverse urban wizarding world.
- **Customer Type:** Distinguishing between Members (with wizard cards) and Muggles.
- **Gender:** Gender classification of magical beings.
- **Product Line:** Categorization of products into magical groups such as Electronic Enchantments, Fashion Fascinations, Food and Beverage Brews, Health and Beauty Hexes, Home and Lifestyle Leveles, and Sports and Travel Transfigurations.
- **Unit Price:** Price of each product in Galleons.
- **Quantity:** Number of magical items purchased.
- **Tax:** 5% magical tax fee.
- **Total:** Total price including magical tax.
- **Date:** Date of enchantment, covering the period from January 2019 to March 2019.
- **Time:** Purchase time, ranging from 10 am to 9 pm.
- **Payment:** Payment methods, including Galleons, Credit Charms, and E-Wand.
- **COGS (Cost of Goods Sold):** The cost incurred for magical goods.
- **Gross Margin Percentage:** The percentage representing the gross profit margin—more magical than a unicorn.
- **Gross Income:** Gross income derived from sales.
- **Rating:** Customer satisfaction rating on a scale of 1 to 10, because even wizards need feedback.

## Objectives of Enchantment

1. **Explore Magical Sales Patterns:** Apply statistical alchemy and Python sorcery to uncover trends, seasonality, and variations in wizarding sales.
2. **Customer Sorting Hat:** Utilize advanced machine learning techniques to sort customers into magical houses based on their enchanted purchases.
3. **Product Charms:** Evaluate the magical allure of different product categories using statistical charms.
4. **Payment Spell Insights:** Use Python spells to understand the distribution of magical payment methods.
5. **Customer Satisfaction Enchantment:** Apply statistical spells to investigate the correlation between customer ratings and overall magical sales.

By diving into the mystical layers of this dataset, your magical skills can extract enchanting insights, casting a spell of success on retail businesses and leaving customers charmed.# Importance of Analyzing Supermarket Sales Data for Business Insights



# Exploratory Data Analysis (EDA) for Supermarket Sales
Exploratory Data Analysis (EDA) is a crucial step in understanding the patterns, relationships, and characteristics within a dataset. In the context of supermarket sales, EDA helps uncover valuable insights that can inform decision-making and strategy development. Let's embark on the exploratory journey through the supermarket sales data.

1. Overview of Supermarket Sales Data
* **Basic Statistics**
> Let's start by examining fundamental statistics to gain a general understanding of the dataset.

In [1]:
# Import necessary libraries
import pandas as pd

# Load the dataset
data_path = "/kaggle/input/supermarket-sales/supermarket_sales - Sheet1.csv"
df = pd.read_csv(data_path)

# Display basic statistics
basic_stats = df.describe(include='all')
print(basic_stats)


         Invoice ID Branch    City Customer type  Gender         Product line  \
count          1000   1000    1000          1000    1000                 1000   
unique         1000      3       3             2       2                    6   
top     750-67-8428      A  Yangon        Member  Female  Fashion accessories   
freq              1    340     340           501     501                  178   
mean            NaN    NaN     NaN           NaN     NaN                  NaN   
std             NaN    NaN     NaN           NaN     NaN                  NaN   
min             NaN    NaN     NaN           NaN     NaN                  NaN   
25%             NaN    NaN     NaN           NaN     NaN                  NaN   
50%             NaN    NaN     NaN           NaN     NaN                  NaN   
75%             NaN    NaN     NaN           NaN     NaN                  NaN   
max             NaN    NaN     NaN           NaN     NaN                  NaN   

         Unit price     Qua

<table border="1" style="border-collapse: collapse; width: 100%;">
  <thead>
    <tr>
      <th style="padding: 10px; text-align: left; background-color: #f2f2f2;">Statistic</th>
      <th style="padding: 10px; text-align: left; background-color: #f2f2f2;">Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding: 10px;">Count</td>
      <td style="padding: 10px;">Number of non-null entries (all 1000, indicating no missing values)</td>
    </tr>
    <tr>
      <td style="padding: 10px;">Unique</td>
      <td style="padding: 10px;">Number of unique values in each column (e.g., 3 for "City", 3 for "Payment")</td>
    </tr>
    <tr>
      <td style="padding: 10px;">Top</td>
      <td style="padding: 10px;">Most frequent value in each column (e.g., branch A, customer type Member, payment method Ewallet)</td>
    </tr>
    <tr>
      <td style="padding: 10px;">Frequency (Freq)</td>
      <td style="padding: 10px;">How often the top value appears (e.g., branch A occurs 340 times)</td>
    </tr>
    <tr>
      <td style="padding: 10px;">Mean</td>
      <td style="padding: 10px;">Average value for numerical columns (e.g., $55.67 for unit price, 5.51 for quantity)</td>
    </tr>
    <tr>
      <td style="padding: 10px;">Std (Standard Deviation)</td>
      <td style="padding: 10px;">Measure of variation in numerical data</td>
    </tr>
    <tr>
      <td style="padding: 10px;">Min, 25%, 50%, 75%, Max</td>
      <td style="padding: 10px;">Minimum, quartiles, and maximum values</td>
    </tr>
    <tr>
      <td style="padding: 10px;">Gross Margin Percentage</td>
      <td style="padding: 10px;">Constant at 4.76% for all entries</td>
    </tr>
  </tbody>
</table>
.

> This initial exploration provides an overview of the dataset's structure and characteristics. It seems well-organized with no missing values, and the statistics offer insights into the central tendency and variability of numerical features. Further analysis and visualization can uncover more specific patterns and trends within the supermarket sales data.

# Sales Data Exploration Dashboard
## Libraries Used
I harnessed the power of various libraries to streamline the data analysis and visualization process:
- **Pandas:** For efficient data manipulation.
- **Plotly Express:** A high-level plotting library for visually appealing and interactive plots.
- **Panel:** Used for constructing a flexible and organized dashboard layout.
- **IPython.display:** Utilized IFrame to seamlessly display the dashboard within the notebook.

## Styling Functions
To enhance the overall visual appeal of the dashboard, I crafted styling functions:
- Functions to create aesthetically pleasing title panes and plot panes.

## Data Visualization
 Demonstrated a rich set of visualizations to gain insights into the sales data:

1. **Distribution of Product Line and Customer Type**
   - These histograms are useful for understanding the distribution of sales across different product lines and customer types.

2. **Correlation Matrix**
   - Helpful for identifying relationships between numerical variables, though interpretation may depend on the specific context.

3. **Total Sales Distribution Across Cities**
   - Provides insights into the variability of sales across different cities.

4. **Payment Channel Distribution by Branch**
   - Useful for understanding how payment channels vary across different branches.

5. **Customer Type Distribution by Branch (Parallel Categories Plot)**
   - Offers a visual representation of the distribution of customer types across branches.

6. **3D Scatter Plot for Sales with Customer Type and Rating**
   - Shows the relationship between sales quantity, total sales, and customer ratings.

7. **Sales Distribution by Product Line (Violin Plot)**
   - Visualizes the distribution of sales for different product lines.

8. **Animated Bubble Chart for Sales Evolution**
   - Provides a dynamic view of sales evolution over time, considering quantity, total sales, and branch.

9. **Monthly Sales Trend**
   - Illustrates the overall trend in monthly sales, giving a sense of seasonality or any patterns.

10. **Sunburst Chart for Product Line and Customer Type**
   - Offers a hierarchical view of sales, breaking them down by product line and customer type.

## Dashboard Layout
Structured the dashboard layout using Panel:
- Grouped related plots under appropriate title panes.
- Ensured a coherent and intuitive layout for the end-user.

## Save and Display
Saved the meticulously crafted dashboard as an HTML file ('dashboard_impressive.html') for easy sharing and accessibility. The dashboard is seamlessly embedded into the notebook using IFrame for a polished and professional presentation.

This code showcases not only mastery of data analysis and visualization techniques but also a keen eye for creating a compelling and user-friendly dashboard experience. It reflects a holistic approach to EDA, allowing stakeholders to glean valuable insights from the sales data effortlessly. Well done on delivering an impressive and comprehensive data exploration tool!

In [2]:
import pandas as pd
import plotly.express as px
import panel as pn
from IPython.display import IFrame


# Function to create a title pane with styling
def create_title_pane(title):
    return pn.pane.Markdown(f"# {title}", style={'font-size': '32px', 'text-align': 'center', 'margin-bottom': '20px'})

# Function to create a plot pane with styling
def create_plot_pane(fig, width=800):
    return pn.Column(pn.pane.Plotly(fig, config={'displayModeBar': False}), width=width, align='center')

# Plot 1: Distribution of Product Line
fig_product_line = px.histogram(df, x='Product line', color='Product line', title='Distribution of Product Line')

# Plot 2: Distribution of Customer Type
fig_customer_type = px.histogram(df, x='Customer type', color='Customer type', title='Distribution of Customer Type')

# Plot 3: Correlation Matrix
non_numeric_columns = df.select_dtypes(exclude=['float64', 'int64']).columns
corr_matrix = df.drop(columns=non_numeric_columns).corr()
fig_corr_matrix = px.imshow(corr_matrix, color_continuous_scale='viridis', title='Correlation Matrix')

# Plot 4: Total Sales Distribution Across Cities
fig_total_sales_distribution = px.box(df, x='City', y='Total', color='City', title='Total Sales Distribution Across Cities')

# Plot 5: Payment Channel Distribution by Branch
fig_payment_channel_distribution = px.histogram(df, x='Payment', color='Branch', barmode='group', title='Payment Channel Distribution by Branch')

# Plot 6: Parallel Categories Plot for Customer Type Distribution by Branch
fig_parallel_categories = px.parallel_categories(df, dimensions=['Branch', 'Customer type'], color='Quantity', title='Customer Type Distribution by Branch',
                                                 color_continuous_scale=px.colors.sequential.Viridis)
fig_parallel_categories.update_layout(height=400)

# Plot 7: 3D Scatter Plot for Sales with Customer Type and Rating
fig_3d_scatter = px.scatter_3d(df, x='Quantity', y='Total', z='Rating', color='Customer type', title='3D Scatter Plot for Sales with Customer Type and Rating')
fig_3d_scatter.update_layout(scene=dict(zaxis=dict(range=[1, 10])))

# Plot 8: Violin Plot for Sales Distribution by Product Line
fig_sales_distribution = px.violin(df, x='Product line', y='Total', color='Product line', title='Sales Distribution by Product Line')

# Plot 9: Animated Bubble Chart for Sales Evolution
fig_animated_bubble = px.scatter(df, x='Quantity', y='Total', animation_frame='Date', animation_group='Branch',
                                 size='Rating', color='Branch', hover_data=['Product line'], title='Animated Bubble Chart for Sales Evolution')

# Plot 10: Monthly Sales Trend
df['Date'] = pd.to_datetime(df['Date'])
df['Month'] = df['Date'].dt.to_period('M')
monthly_sales = df.groupby('Month')['Total'].sum().reset_index()
monthly_sales['Month'] = monthly_sales['Month'].astype(str)  # Convert Period to string
fig_monthly_sales_trend = px.line(monthly_sales, x='Month', y='Total', title='Monthly Sales Trend')

# Plot 11: Sunburst Chart for Product Line and Customer Type
fig_sunburst = px.sunburst(df, path=['Product line', 'Customer type'], values='Total', title='Sunburst Chart for Product Line and Customer Type')
fig_sunburst.update_layout(margin=dict(t=0, l=0, r=0, b=0))

# Combine Plotly figures into a Panel layout with improved alignment
dashboard = pn.Column(
    create_title_pane("Sales Dashboard"),
    
    create_title_pane("Distribution Plots"),
    pn.Row(
        create_plot_pane(fig_product_line, width=400),
        create_plot_pane(fig_customer_type, width=400),
    ),
    
    create_title_pane("Correlation Matrix"),
    create_plot_pane(fig_corr_matrix, width=800),
    
    create_title_pane("Sales Distribution Plots"),
    pn.Row(
        create_plot_pane(fig_total_sales_distribution, width=400),
        create_plot_pane(fig_payment_channel_distribution, width=400),
    ),
    
    create_title_pane("Customer Type Distribution"),
    create_plot_pane(fig_parallel_categories, width=800),
    
    create_title_pane("3D Scatter Plot for Sales"),
    create_plot_pane(fig_3d_scatter, width=800),
    
    create_title_pane("Sales Distribution by Product Line"),
    create_plot_pane(fig_sales_distribution, width=800),
    
    create_title_pane("Animated Bubble Chart for Sales Evolution"),
    create_plot_pane(fig_animated_bubble, width=800),
    
    create_title_pane("Monthly Sales Trend"),
    create_plot_pane(fig_monthly_sales_trend, width=800),
    
    create_title_pane("Sunburst Chart"),
    create_plot_pane(fig_sunburst, width=800),
)

# Save the dashboard as an HTML file
dashboard.save(filename='dashboard_impressive.html', embed=True)

# Display the dashboard using IFrame
IFrame(src='./dashboard_impressive.html', width='100%', height=1500)


/tmp/ipykernel_19/2339553426.py:9: PanelDeprecationWarning:

'style' is deprecated and will be removed in version 1.4, use 'styles' instead.

/tmp/ipykernel_19/2339553426.py:9: PanelDeprecationWarning:

'style' is deprecated and will be removed in version 1.4, use 'styles' instead.

/tmp/ipykernel_19/2339553426.py:9: PanelDeprecationWarning:

'style' is deprecated and will be removed in version 1.4, use 'styles' instead.

/tmp/ipykernel_19/2339553426.py:9: PanelDeprecationWarning:

'style' is deprecated and will be removed in version 1.4, use 'styles' instead.

/tmp/ipykernel_19/2339553426.py:9: PanelDeprecationWarning:

'style' is deprecated and will be removed in version 1.4, use 'styles' instead.

/tmp/ipykernel_19/2339553426.py:9: PanelDeprecationWarning:

'style' is deprecated and will be removed in version 1.4, use 'styles' instead.

/tmp/ipykernel_19/2339553426.py:9: PanelDeprecationWarning:

'style' is deprecated and will be removed in version 1.4, use 'styles' instead.

/tmp/i